keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via numeric parameters, then NumericParameters must list all claim
+    // parameter types that are allowed for claims.
+    NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResourceCapacity
+
+For each node, a NodeResourceCapacity objects gets created:
+
+```
+// NodeResourceCapacity gets published by kubelet. Its name
+// matches the name of the node and the node is the owner.
+//
+// Capacity may get added, but should not get removed because
+// it would make scheduling decisions based on the old capacity
+// invalid.
+type NodeResourceCapacity struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Resources contains information about the capacity
+	// reported by each DRA driver.
+	//
+	// +optional
+	Resources []DriverResources
+}
+
+type DriverResources struct {
+	// DriverName identifies the DRA which provided the capacity information.
+	DriverName string
+
+	// ResourceInstances describes all discrete resource instances that are
+	// managed by the driver. Each entry must be an object of one of the
+	// supported numeric resource capacity types with kind, version and uid
+	// set. The uid only needs to be unique inside the node.
+	//
+	// +optional
+	ResourceInstances []runtime.RawExtension
+}
+```
+
+If a driver needs to remove capacity, then there is a risk that a claim gets
+allocated using that capacity while the kubelet is updating the
+NodeResourceCapacity object. The implementations of numeric models must handle
+scenarios where more resources are allocated than available. The DRA driver
+kubelet plugin must double-check that the allocated resources are still
+available when NodePrepareResource is called. If not, the pod cannot cannot
+start until the resource comes back.
```

Comment:
Or we have the ability to bounce failed pods back to the scheduler under certain conditions: https://kubernetes.slack.com/archives/C5P3FE08M/p1705584616712299

Reply:
This can also be used above, if the CR generation has changed by the time the allocation reaches the driver.

Reply:
Note: this is an existing issue with the traditional device plugin API, today (which is exacerbated by the existence of the TopologyManager).

I don't think we should assume support for such a "bounce-back" capability is a prerequisite for this KEP, but it will improve the story around such transient failures significantly.